### DISCUSSION TITLE
Adding mail support

### DISCUSSION THREAD

**Discussion by aidanheerdegen** [payu-org/payu#295] [Ideas]
Chatting with @dkhutch about the changes he made to his fork of payu he used in Sweden

https://github.com/dkhutch/payu/tree/tetralith

he added support for mailing the user on error, or when a job (or sequence of jobs) finished.

I think this is a great idea.

**Comment by dkhutch** [comment]
Following on from this… I would suggest the following tweak to the current version of payu:
qsub has a flag: -m which sends mail to the user. The one I’m interested in is:
qsub -m a <script> 
which sends the user an email when the job terminates. 
I would like to edit payu so that:
if n_runs == 1:
   <add -m a option to pbs_flags> 
That way when a job reaches its last iteration, I will receive an email when the job completes.

**Reply by aidanheerdegen** [reply]
From the `qsub` man page:
```
-m <mail events>
               Specifies  the  set  of  conditions under which mail about the job is sent.  Sets job's Mail_Points attribute to mail events.  The mail events
               argument can be one of the following:
                  The single character "n"
                  Any combination of "a", "b", and "e", with optional "j"

               The following table lists the sub-options to the -m option:

               n    No mail is sent

               a    Mail is sent when the job is aborted by the batch system

               b    Mail is sent when the job begins execution

               e    Mail is sent when the job terminates

               j    Mail is sent for subjobs. Must be combined with one or more of the a , b , or e sub-options

               Format: String
               Syntax: n | [j](one or more of a, b, e)
               Example: -m ja
               Default: a
```
So I think the option you want is `-m e` for `n_runs == 1` and `-m a` otherwise?

There should be some method to turn this on or off in `config.yaml`.

One possibility:
```yaml
mail: true
```
for mailing on error and at the end of a run and
```yaml
mail:
   error: true
   finish: false
```
if you wanted to break it down? Maybe that isn't necessary.

**Reply by dkhutch** [reply]
Oh yes sorry! `-m a` is the default so I would want `-m e` for n_runs == 1

Yes I suppose if other users don't want this notification you can build an option to switch it on and off. (I always have it on in my Swedish version.)

**Reply by aidanheerdegen** [reply]
The man page says `-m a` is the default, but I don't recall getting messages when jobs failed. 

I did some testing, and sent a query to NCI, turns out no messages get through without setting an explicit email address with the `-M` option.

Even after using the `-M` option no email is sent on failure by default either. In fact the `-a` option doesn't seem to be triggered when a job fails with an exit status > 0, or even killed for exceeding PBS limits like walltime. I am not sure what "aborted by the batch system" even means in that case.

Well I asked NCI and they said abort was an internal PBS state, and not related to the exit status of the job. So we don't have the ability to differentiate between a successful or unsuccessful job. That is problematic.

**Reply by dkhutch** [reply]
Thanks for looking into it... that is frustrating.

**Reply by aidanheerdegen** [reply]
Within `payu` we could submit a job whose sole purpose is tell you a job has failed. That is a bit icky though.

**Reply by dkhutch** [reply]
Yeah that's basically how I implemented my 'job is finished' notification. I basically had `if n_runs == 0: > create mail job; submit mail job`

**Reply by aidanheerdegen** [reply]
We don't have to do it that way for the "job is finished" notification, can check if `n_runs == 0` and then add the `-m e` flag to the last job. That is clean and will work for both schedulers I guess.

**Reply by dkhutch** [reply]
Yep that is definitely cleaner for that part of it - though you'd need to do it at `n_runs == 1` in that case.

**Reply by aidanheerdegen** [reply]
Yes, did mean to change that to `n_runs == 1`. 

An email on error can be generated by submitting a job whose sole purpose is to convey the information that the calling job has failed. This is clumsy, will have to wait until the queued job runs and has limited ability to pass relevant information to the email sent. In this case it is only the job name that appears in the email, and that has limitations (no spaces or special characters). 

The body of such an email might look something like this:
```
PBS Job Id: 17394310.gadi-pbs
Job Name:   payu_error_jobname
Execution terminated
Exit_status=0
resources_used.cpupercent=0
resources_used.cput=00:00:00
resources_used.mem=6764kb
resources_used.ncpus=1
resources_used.vmem=6764kb
resources_used.walltime=00:00:01
```

**Reply by aidanheerdegen** [reply]
There is also the option of using `mailx` on `accessdev` like so:
```
ssh accessdev 'echo "something" | mailx -s "subject" aph502'
```
but this requires either passwordless `ssh` keys or forwarding an authentication agent to the job. Seems complicated.

**Reply by marshallward** [reply]
This mail query would only be done for PBS, right?

**Reply by dkhutch** [reply]
In response to Aidan's suggestion to create a `payu_error_jobname` script. It might seem clumsy, but I would certainly find it useful. I would simply use the original payu job name as an identifier. When the user receives this email, they will then have to log on and check the run to see what actually happened. But that's what you need... a prompt to log on and check that particular run that has crashed.

**Reply by aidanheerdegen** [reply]
Yeah the idea would be to make it something like `payu_error_$jobname`, e.g. `payu_error_run_1deg_iaf`.

It is good to know that even this limited functionality would be useful.

**Comment by aidanheerdegen** [comment]
ping @marshallward in case he didn't see this discussion

**Reply by marshallward** [reply]
I've been following, sounds like a good idea.

I think slurm has a similar feature so might be a good idea to add it to the scheduler abstraction.

**Reply by aidanheerdegen** [reply]
In fact David's fork was also modified to use `slurm`, and why he wanted to add this to the system `payu`. 

It is also useful to check the `slurm` scheduler support.

**Reply by dkhutch** [reply]
The slurm system I used in Sweden had a mail option called 'FAIL', which is pretty handy. Unfortunately the PBS system doesn't seem to have an equivalent.

**Reply by aidanheerdegen** [reply]
Yeah, and that is what is annoying/difficult about supporting multiple schedulers, allowing for the overlap, and lack thereof.

**Reply by marshallward** [reply]
`raise NotImplementedError` in PBS?

**Reply by aidanheerdegen** [reply]
Will require good documentation, love good docs, hate writing them.
